src/Combinatorics/EnumerativeCombinatorics/multipartitions.jl

```diff
@@ -0,0 +1,177 @@
+################################################################################
+# Multipartitions.
+#
+# Copyright (C) 2020 Ulrich Thiel, ulthiel.com/math
+#
+# Originally taken from the JuLie [repository](https://github.com/ulthiel/JuLie)
+# by Ulrich Thiel and OSCAR-ified by Morgan Rodgers.
+################################################################################
+
+
+
+
+"""
+    multipartition(mp::Vector{Partition{T}}) where T <: IntegerUnion
+    multipartition(mp::Vector{Vector{T}}) where T <: IntegerUnion
+
+Return the multipartition given by the vector `mp` of integer sequences `mp`
+(which are interpreted as integer partitions) as an object of type
+`Multipartition{T}`.
+Note that a partition must be given by a weakly descending integer sequence.
+
+The element type `T` may be optionally specified, see also the examples below.
+
+# Examples
+```jldoctest
+julia> P = multipartition([[2,1], [], [3,2,1]])
+Partition{Int64}[[2, 1], [], [3, 2, 1]]
+
+julia> sum(P)
+9
+
+julia> P[2]
+Empty partition
+
+julia> P = multipartition(Vector{Int8}[[2,1], [], [3,2,1]])
+Partition{Int8}[[2, 1], [], [3, 2, 1]]
+```
+"""
+function multipartition(mp::Vector{Partition{T}}) where T <: IntegerUnion
+  return Multipartition(mp)
+end
+
+function multipartition(mp::Vector{Vector{T}}) where T <: IntegerUnion
+  return Multipartition([partition(p) for p in mp])
+end
+
+# This is only called when the empty array is part of mp (because then it's
+# "Any" type and not of Integer type).
+function multipartition(mp::Vector{Vector{Any}})
+  return Multipartition([partition(p) for p in mp])
+end
+
+# This is only called when mp is itself an empty array.
+function multipartition(mp::Vector{Any})
+  return Multipartition(Vector{Partition{Int}}(mp))
+end
+
+function Base.show(io::IO, ::MIME"text/plain", MP::Multipartition)
+  print(io, MP.mp)
+end
+
+function Base.size(MP::Multipartition)
+  return size(MP.mp)
+end
+
+function Base.length(MP::Multipartition)
+  return length(MP.mp)
+end
+
+function Base.getindex(MP::Multipartition, i::Int)
+  return getindex(MP.mp,i)
+end
+
+
+"""
+    sum(P::Multipartition{T}) where T<:IntegerUnion
+
+If `P` is a multipartition of the integer n, this function returns n.
+"""
+function Base.sum(MP::Multipartition{T}) where T<:IntegerUnion
+  return sum(sum, MP.mp; init=zero(T))
+end
+
+
+"""
+    multipartitions(n::T, r::IntegerUnion)  where T<:IntegerUnion
+
+A list of all `r`-component multipartitions of `n`, as elements of type Multipartitoon{T}.
+The algorithm is recursive and  based on [`partitions(::IntegerUnion)`](@ref).
+
+# Example
+```jldoctest
+julia> MP = multipartitions(2,2);
+
+julia> first(MP)
+Partition{Int64}[[], [2]]
+
+julia> collect(multipartitions(2,2))
+5-element Vector{Multipartition{Int64}}:
```

Comment:
Is this perhaps like this?
```suggestion
5-element Vector{Partition{Int64}}:
```

Reply:
But CI test pass... It seems I am misunderstanding something... huh

Reply:
I agree that it seems like this should be `Vector{Multipartition}`, this definitely passes CI but that doesn't mean that it's what we want... I'm looking at it now.

Reply:
Ah okay, I understand the confusion. `multipartitions` returns ALL the multipartitions of n into r parts, so we get a `Vector{Multipartition}`. While each individual `Multipartion` looks like a `Vector{Partition}`. So this is all correct.

Reply:
Ahah I get it now:
```
julia> foo = collect(MP)
5-element Vector{Multipartition{Int64}}:
 Partition{Int64}[[], [2]]
 Partition{Int64}[[], [1, 1]]
 Partition{Int64}[[1], [1]]
 Partition{Int64}[[2], []]
 Partition{Int64}[[1, 1], []]

julia> typeof(foo)
Vector{Multipartition{Int64}} (alias for Array{Multipartition{Int64}, 1})

julia> foo[1]
Partition{Int64}[[], [2]]

julia> typeof(foo[1])
Multipartition{Int64}
```
So my confusion is that `Multipartition` is printed as a Vector{Partition}`. Which makes some sense, but at least for me it would be easier to understand what is going on if there was a custom `show` method...

Reply:
I can do this but I'm not sure what it should look like. We have the same issue with `Partition` printing as `Vector{Int}`:
```julia-repl
julia> P = partition(6, 4, 4, 2)
[6, 4, 4, 2]
```